Zadanie 1

a) Ściągnij plik iris.csv i umieść go w katalogu z plikami pythonowymi.
b) Użyj paczki pandas do wczytania tego pliku do Pythona i wyświetl bazę danych.
c) W obiekcie df mamy bazę danych typu dataframe. Żeby bezpośrednio wejść
do danych z tej bazy trzeba odwołać się do wartości (values).

In [217]:
import pandas as pd

df2 = pd.read_csv("diabetes.csv")

print(df2.values)

[[6 148 72 ... 0.627 50 'tested_positive']
 [1 85 66 ... 0.351 31 'tested_negative']
 [8 183 64 ... 0.672 32 'tested_positive']
 ...
 [5 121 72 ... 0.245 30 'tested_negative']
 [1 126 60 ... 0.349 47 'tested_positive']
 [1 93 70 ... 0.315 23 'tested_negative']]


d) Spróbujmy wyświetlić różne fragmenty danych.

In [95]:
#wszystkie wiersze, kolumna nr 0
print(df.values[:, 0])
#wiersze od 5 do 10, wszystkie kolumny
print(df.values[5:11, :])
#dane w komórce [1,4]
print(df.values[1, 4])

[5.1 4.9 4.7 4.6 5.0 5.4 4.6 5.0 4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.0 5.0 5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.0
 5.5 4.9 4.4 5.1 5.0 4.5 4.4 5.0 5.1 4.8 5.1 4.6 5.3 5.0 7.0 6.4 6.9 5.5
 6.5 5.7 6.3 4.9 6.6 5.2 5.0 5.9 6.0 6.1 5.6 6.7 5.6 5.8 6.2 5.6 5.9 6.1
 6.3 6.1 6.4 6.6 6.8 6.7 6.0 5.7 5.5 5.5 5.8 6.0 5.4 6.0 6.7 6.3 5.6 5.5
 5.5 6.1 5.8 5.0 5.6 5.7 5.7 6.2 5.1 5.7 6.3 5.8 7.1 6.3 6.5 7.6 4.9 7.3
 6.7 7.2 6.5 6.4 6.8 5.7 5.8 6.4 6.5 7.7 7.7 6.0 6.9 5.6 7.7 6.3 6.7 7.2
 6.2 6.1 6.4 7.2 7.4 7.9 6.4 6.3 6.1 7.7 6.3 6.4 6.0 6.9 6.7 6.9 5.8 6.8
 6.7 6.7 6.3 6.5 6.2 5.9]
[[5.4 3.9 1.7 0.4 'Setosa']
 [4.6 3.4 1.4 0.3 'Setosa']
 [5.0 3.4 1.5 0.2 'Setosa']
 [4.4 2.9 1.4 0.2 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [5.4 3.7 1.5 0.2 'Setosa']]
Setosa


e) Pora na podział zbioru na zbiór testowy i treningowy. Wykorzystamy do tego
funkcję train_test_split z paczki sklearn. Dodaj na górze programu odpowiedni

In [96]:
from sklearn.model_selection import train_test_split

Następnie dopisz na dole programu funckję dzielącą zbiór df.values na dwa
kawałki w proporcjach 70% i 30%, w losowy sposób.

In [97]:
#podzial na zbior testowy (30%) i treningowy (70%), ziarno losowosci = 13
(train_set, test_set) = train_test_split(df.values, train_size=0.7,
                                         random_state=274929)

f) Sprawdź jakie rekordy znalazły się w zbiorze testowym i ile ma on rekordów.

In [98]:
print(test_set)
print(test_set.shape[0])

[[5.7 2.8 4.5 1.3 'Versicolor']
 [5.6 2.9 3.6 1.3 'Versicolor']
 [5.1 3.7 1.5 0.4 'Setosa']
 [5.8 2.7 5.1 1.9 'Virginica']
 [6.8 3.0 5.5 2.1 'Virginica']
 [5.5 4.2 1.4 0.2 'Setosa']
 [7.7 3.0 6.1 2.3 'Virginica']
 [6.3 2.7 4.9 1.8 'Virginica']
 [5.1 3.8 1.6 0.2 'Setosa']
 [5.5 2.6 4.4 1.2 'Versicolor']
 [6.0 3.0 4.8 1.8 'Virginica']
 [6.3 2.8 5.1 1.5 'Virginica']
 [6.3 2.3 4.4 1.3 'Versicolor']
 [5.1 3.8 1.9 0.4 'Setosa']
 [5.8 2.8 5.1 2.4 'Virginica']
 [4.8 3.1 1.6 0.2 'Setosa']
 [6.1 3.0 4.9 1.8 'Virginica']
 [6.5 3.0 5.8 2.2 'Virginica']
 [6.1 3.0 4.6 1.4 'Versicolor']
 [4.4 2.9 1.4 0.2 'Setosa']
 [6.0 2.9 4.5 1.5 'Versicolor']
 [5.7 4.4 1.5 0.4 'Setosa']
 [5.1 3.5 1.4 0.3 'Setosa']
 [7.4 2.8 6.1 1.9 'Virginica']
 [5.4 3.0 4.5 1.5 'Versicolor']
 [6.7 3.3 5.7 2.5 'Virginica']
 [5.8 4.0 1.2 0.2 'Setosa']
 [7.7 2.6 6.9 2.3 'Virginica']
 [6.1 2.9 4.7 1.4 'Versicolor']
 [6.4 2.8 5.6 2.2 'Virginica']
 [6.4 3.1 5.5 1.8 'Virginica']
 [5.2 4.1 1.5 0.1 'Setosa']
 [5.4 3.4 1.5 0.4 'Setosa']
 [

g) Bywa, że zbiór testowy i treningowy trzeba podzielić dodatkowo na inputy
(kolumny numeryczne) i klasy (odmiany irysów). Wówczas z 2 zbiorów danych
zrobią się nam cztery.

In [99]:
train_inputs = train_set[:, 0:4]
train_classes = train_set[:, 4]
test_inputs = test_set[:, 0:4]
test_classes = test_set[:, 4]

Zadanie 2

a) Rozpocznij od przygotowania zbioru treningowego i testowego, tak jak w
poprzednim zadaniu.

In [100]:
import pandas as pd
from sklearn.model_selection import train_test_split

df = pd.read_csv("iris.csv")
(train_set, test_set) = train_test_split(df.values, train_size=0.7,
                                         random_state=274929)

b) Napiszemy w Pythonie funkcję, która na podstawie czterech numerycznych
parametrów irysa odgadnie jego gatunek wykorzystując do tego jedynie dwie
instrukcje if, elif oraz else (jednokrotnie).

In [117]:
def classify_iris(sl, sw, pl, pw):
    if sl > 4:
        return("Setosa")
    elif pl <= 5:
        return("Virginica")
    else:
        return("Versicolor")

c)

In [192]:
good_predictions = 0
len = test_set.shape[0]

for i in range(len):
    if classify_iris(test_inputs[i][0], test_inputs[i][1], test_inputs[i][2], test_inputs[i][3]) == test_set[i][4]:
        good_predictions += 1

print(good_predictions)
print(good_predictions/len*100, "%")

17
37.77777777777778 %


Wyniki:
17
37.77777777777778 %

d) Poprawiony klasyfikator.

In [211]:
def classify_iris2(sl, sw, pl, pw):
    if pw < 1:
        return("Setosa")
    elif pl > 4.8 and pw > 1.7:
        return("Virginica")
    else:
        return("Versicolor")

In [172]:
#sorted
print(train_set[train_set[:, 4].argsort()])

[[5.0 3.6 1.4 0.2 'Setosa']
 [5.2 4.1 1.5 0.1 'Setosa']
 [5.0 3.5 1.6 0.6 'Setosa']
 [4.8 3.0 1.4 0.3 'Setosa']
 [5.5 4.2 1.4 0.2 'Setosa']
 [5.1 3.8 1.9 0.4 'Setosa']
 [5.1 3.5 1.4 0.2 'Setosa']
 [5.1 3.8 1.6 0.2 'Setosa']
 [4.8 3.0 1.4 0.1 'Setosa']
 [4.6 3.2 1.4 0.2 'Setosa']
 [4.8 3.1 1.6 0.2 'Setosa']
 [5.4 3.4 1.7 0.2 'Setosa']
 [5.0 3.2 1.2 0.2 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [5.4 3.9 1.7 0.4 'Setosa']
 [5.2 3.4 1.4 0.2 'Setosa']
 [5.1 3.7 1.5 0.4 'Setosa']
 [5.7 4.4 1.5 0.4 'Setosa']
 [4.4 3.2 1.3 0.2 'Setosa']
 [5.4 3.4 1.5 0.4 'Setosa']
 [5.1 3.3 1.7 0.5 'Setosa']
 [5.8 4.0 1.2 0.2 'Setosa']
 [4.9 3.0 1.4 0.2 'Setosa']
 [5.0 3.3 1.4 0.2 'Setosa']
 [5.0 3.0 1.6 0.2 'Setosa']
 [4.3 3.0 1.1 0.1 'Setosa']
 [5.3 3.7 1.5 0.2 'Setosa']
 [5.2 3.5 1.5 0.2 'Setosa']
 [5.4 3.9 1.3 0.4 'Setosa']
 [5.1 3.8 1.5 0.3 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [4.6 3.6 1.0 0.2 'Setosa']
 [6.0 2.7 5.1 1.6 'Versicolor']
 [6.7 3.1 4.4 1.4 'Versicolor']
 [6.0 2.2 4.

In [212]:
good_predictions2 = 0
len = test_set.shape[0]

for i in range(len):
    if classify_iris2(test_inputs[i][0], test_inputs[i][1], test_inputs[i][2], test_inputs[i][3]) == test_set[i][4]:
        good_predictions2 += 1

print(good_predictions2)
print(good_predictions2/len*100, "%")

42
93.33333333333333 %


Wyniki:
42
93.3%

zadanie 3

a) Podziel w losowy sposób bazę danych irysów na zbiór treningowy i zbiór
testowy w proporcjach 70%/30%. Wyświetl oba zbiory. Podziel te zbiory na
cztery części (inputy i class), jeśli jest taka potrzeba.

In [156]:
df = pd.read_csv("iris.csv")
(train_set, test_set) = train_test_split(df.values, train_size=0.7,
                                         random_state=274929)

train_inputs = train_set[:, 0:4]
train_classes = train_set[:, 4]
test_inputs = test_set[:, 0:4]
test_classes = test_set[:, 4]

b) Zainicjuj drzewo decyzyjne metodą DecisionTreeClassifier.

In [157]:
from sklearn.tree import DecisionTreeClassifier

dtc = DecisionTreeClassifier()

c) Wytrenuj drzewo decyzyjne na zbiorze treningowym, wykorzystując funkcję fit.

In [213]:
dtc.fit(train_inputs, train_classes)

DecisionTreeClassifier()

d) Wyświetl drzewo w formie tekstowej i/lub w formie graficznej. Jeśli masz
problemy z tym podpunktem – pomiń go.

In [214]:
from sklearn.tree import export_text

feature_names = ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

tree_to_print = export_text(dtc, feature_names=feature_names)
print(tree_to_print)

|--- petal width (cm) <= 0.80
|   |--- class: Setosa
|--- petal width (cm) >  0.80
|   |--- petal length (cm) <= 4.85
|   |   |--- petal width (cm) <= 1.70
|   |   |   |--- class: Versicolor
|   |   |--- petal width (cm) >  1.70
|   |   |   |--- sepal width (cm) <= 3.00
|   |   |   |   |--- class: Virginica
|   |   |   |--- sepal width (cm) >  3.00
|   |   |   |   |--- class: Versicolor
|   |--- petal length (cm) >  4.85
|   |   |--- petal width (cm) <= 1.70
|   |   |   |--- petal width (cm) <= 1.55
|   |   |   |   |--- class: Virginica
|   |   |   |--- petal width (cm) >  1.55
|   |   |   |   |--- petal length (cm) <= 5.45
|   |   |   |   |   |--- class: Versicolor
|   |   |   |   |--- petal length (cm) >  5.45
|   |   |   |   |   |--- class: Virginica
|   |   |--- petal width (cm) >  1.70
|   |   |   |--- class: Virginica



e) Dokonaj ewaluacji klasyfikatora: sprawdź jak drzewo poradzi sobie z rekordami
ze zbioru testowego. Wyświetl dokładność klasyfikatora, czyli procent
poprawnych odpowiedzi. Wykorzystaj do tego funkcję score lub predict.

In [166]:
test_score = dtc.score(test_inputs, test_classes)
print(test_score)
test_predict = dtc.predict(test_inputs)
print(test_predict)

0.9111111111111111
['Versicolor' 'Virginica' 'Setosa' 'Virginica' 'Setosa' 'Versicolor'
 'Versicolor' 'Virginica' 'Virginica' 'Setosa' 'Setosa' 'Versicolor'
 'Setosa' 'Setosa' 'Virginica' 'Versicolor' 'Virginica' 'Virginica'
 'Virginica' 'Virginica' 'Virginica' 'Virginica' 'Versicolor' 'Versicolor'
 'Virginica' 'Setosa' 'Versicolor' 'Versicolor' 'Setosa' 'Virginica'
 'Versicolor' 'Setosa' 'Setosa' 'Virginica' 'Versicolor' 'Setosa' 'Setosa'
 'Setosa' 'Setosa' 'Setosa' 'Versicolor' 'Setosa' 'Setosa' 'Virginica'
 'Versicolor']


f) Wyświetl macierz błędów (ang. confusion matrix), która zestawia liczby
błędnych i poprawnych odpowiedzi, dla wszystkich klas.

In [215]:
from sklearn.metrics import confusion_matrix

confusion_matrix(test_classes, test_predict, labels=["Setosa", "Virginica", "Versicolor"])

array([[17,  0,  0],
       [ 0, 12,  1],
       [ 0,  3, 12]], dtype=int64)

Czy drzewo nauczone automatycznie jest lepsze niż Twoje drzewko stworzone ręcznie
w zadaniu 2?
Tak, Tutaj udało się osiągnąć 97.7%.
Ja osiągnąłem 93.3%.

Zadanie 4

a), b), c)

In [ ]:
import pandas as pd

df = pd.read_csv("iris.csv")

print(df.values)